core/src/main/scala/org/apache/spark/storage/MemoryStore.scala

```diff
@@ -235,6 +246,43 @@ private[spark] class MemoryStore(blockManager: BlockManager, memoryManager: Memo
   }
 
   /**
+   * [[unrollSafely]] is modified to detect over-sized block,
+   * but it might give false negative when not enough storage memory is available.
+   *
+   * In this case, this api is used by CSD as a post process of [[unrollSafely]] to detect
+   * partitions larger than 2G in estimated object size.
+   *
+   * Since this api fetches until we "see" 2G of object in size, we need
+   * sufficient memory in spark's user memory space to be set apart:
+   * (2G + overhead) per thread and per operator/RDD compute.
+   */
+  def fetchUntilCsdBlockSizeLimit[T](
```

Comment:
This api is challenging as we need to keep all the seen values until reaching 2G size limit. it is a potential OOM thread and hard to test.